api/types/registry/registry.go

```diff
@@ -117,3 +121,143 @@ type DistributionInspect struct {
 	// obtained by parsing the manifest
 	Platforms []v1.Platform
 }
+
+// Registry holds information for a registry and its mirrors.
+type Registry struct {
+	// Prefix is used for the lookup of endpoints, where the given registry
+	// is selected when its Prefix is a prefix of the passed reference, for
+	// instance, Prefix:"docker.io/opensuse" will match a `docker pull
+	// opensuse:tumleweed`.
+	URL RegURL `json:"Prefix"`
+	// The mirrors will be selected prior to the registry during lookup of
+	// endpoints.
+	Mirrors []Mirror `json:"Mirrors,omitempty"`
+}
+
+// NewRegistry returns a Registry and interprets input as a URL.
+func NewRegistry(input string) (Registry, error) {
+	reg := Registry{}
+	err := reg.URL.Parse(input)
+	return reg, err
+}
+
+// AddMirror interprets input as a URL and adds it as a new mirror.
+func (r *Registry) AddMirror(input string) error {
+	mir, err := NewMirror(input)
+	if err != nil {
+		return err
+	}
+	r.Mirrors = append(r.Mirrors, mir)
+	return nil
+}
+
+// ContainsMirror returns true if the URL of any mirror equals input.
+func (r *Registry) ContainsMirror(input string) bool {
+	for _, m := range r.Mirrors {
+		if m.URL.String() == input {
+			return true
+		}
+	}
+	return false
+}
+
+// Mirror holds information for a given registry mirror.
+type Mirror struct {
+	// The URL of the mirror.
+	URL RegURL `json:"URL,omitempty"`
```

Comment:
Could you replace `URL RegURL` with `Address string`, and put `Net string` field to allow using UNIX socket? 

e.g. `{"Address": "/run/docker-registry-cache.sock", "Net": "unix"}`, `{"Address": "https://example.com", "Net": "tcp"}`

Implementation of UNIX sockets itself can be done in a separate PR. (https://github.com/moby/moby/pull/36310)

cc @cpuguy83

Reply:
@AkihiroSuda Currently, I am hesitating to add (yet another) feature to this PR. It is already very contentious, has been rewritten multiple times and I'd like to avoid that we have to celebrate a birthday for this PR :-D

The data types are designed in a way that they can be extended in the future, so I would really appreciate if we could defer the socket support and keep the discussion on how to design that in #36310. There's also the wildcard support on the todo list.

Reply:
No need to implement this new feature.  I will do.
Could you just consider s/url/addr/ for extensibility

Reply:
> No need to implement this new feature. I will do.

Thanks!

> Could you just consider s/url/addr/ for extensibility

I am fine with renaming that (Cc @thaJeztah, @stevvooe). Shall we change the type to `string` as well?

Reply:
@vrothberg please. :)